File: go/test/department_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/met-museum-collection-sdk"
	"github.com/voxgig-sdk/met-museum-collection-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestDepartmentEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Department(nil)
		if ent == nil {
			t.Fatal("expected non-nil DepartmentEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := departmentBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "department." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set METMUSEUMCOLLECTION_TEST_DEPARTMENT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		departmentRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.department", setup.data)))
		var departmentRef01Data map[string]any
		if len(departmentRef01DataRaw) > 0 {
			departmentRef01Data = core.ToMapAny(departmentRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = departmentRef01Data

		// LIST
		departmentRef01Ent := client.Department(nil)
		departmentRef01Match := map[string]any{}

		departmentRef01ListResult, err := departmentRef01Ent.List(departmentRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, departmentRef01ListOk := departmentRef01ListResult.([]any)
		if !departmentRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", departmentRef01ListResult)
		}

	})
}

func departmentBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "department", "DepartmentTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read department test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse department test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"department01", "department02", "department03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("METMUSEUMCOLLECTION_TEST_DEPARTMENT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"METMUSEUMCOLLECTION_TEST_DEPARTMENT_ENTID": idmap,
		"METMUSEUMCOLLECTION_TEST_LIVE":      "FALSE",
		"METMUSEUMCOLLECTION_TEST_EXPLAIN":   "FALSE",
		"METMUSEUMCOLLECTION_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["METMUSEUMCOLLECTION_TEST_DEPARTMENT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["METMUSEUMCOLLECTION_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["METMUSEUMCOLLECTION_APIKEY"],
			},
			extra,
		})
		client = sdk.NewMetMuseumCollectionSDK(core.ToMapAny(mergedOpts))
	}

	live := env["METMUSEUMCOLLECTION_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["METMUSEUMCOLLECTION_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/department_entity_test.rb
# Department entity test

require "minitest/autorun"
require "json"
require_relative "../MetMuseumCollection_sdk"
require_relative "runner"

class DepartmentEntityTest < Minitest::Test
  def test_create_instance
    testsdk = MetMuseumCollectionSDK.test(nil, nil)
    ent = testsdk.Department(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = department_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "department." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set METMUSEUMCOLLECTION_TEST_DEPARTMENT_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    department_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.department")))
    department_ref01_data = nil
    if department_ref01_data_raw.length > 0
      department_ref01_data = Helpers.to_map(department_ref01_data_raw[0][1])
    end

    # LIST
    department_ref01_ent = client.Department(nil)
    department_ref01_match = {}

    department_ref01_list_result, err = department_ref01_ent.list(department_ref01_match, nil)
    assert_nil err
    assert department_ref01_list_result.is_a?(Array)

  end
end

def department_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "department", "DepartmentTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = MetMuseumCollectionSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["department01", "department02", "department03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["METMUSEUMCOLLECTION_TEST_DEPARTMENT_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "METMUSEUMCOLLECTION_TEST_DEPARTMENT_ENTID" => idmap,
    "METMUSEUMCOLLECTION_TEST_LIVE" => "FALSE",
    "METMUSEUMCOLLECTION_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["METMUSEUMCOLLECTION_TEST_DEPARTMENT_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["METMUSEUMCOLLECTION_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = MetMuseumCollectionSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["METMUSEUMCOLLECTION_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["METMUSEUMCOLLECTION_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
